Add mani cli tool

thornycrackers opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Add the mani cli tool

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [*] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).


I don't really know much GO so I mostly copied and pasted. The tests were trying to create files and got permission denied so I just skipped them for now. The resulting binary seems to be working fine, haven't had any issues yet.

teto commented:
I've been wanting such a tool for some time (I think it fits the bill). Would be nice to have autocompletion scripts installed too. (grep nixpkgs for installShellFile)

legendofmiracles commented on pkgs/development/tools/mani/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  vendorSha256 = "sha256-ZivzDfjx2djzS0Xm3GISK3zpB5fUUMgy2o4Ti1Z9wMM=";
+
+  # Skip tests
+  checkPhase = null;
```
```suggestion
  # Skip tests
  doCheck = false;
```
Can you please add a comment why we are skipping the checks?

thornycrackers replied:
Updated

legendofmiracles commented on pkgs/development/tools/mani/default.nix:
```diff
@@ -0,0 +1,32 @@
+    homepage = "https://manicli.com/";
+    changelog = "https://github.com/alajmo/mani/releases/tag/v${version}";
+    license = licenses.mit;
+    platforms = platforms.unix;
```
```suggestion
```
Set by buildGoModule

thornycrackers replied:
updated

legendofmiracles commented:
This should also depend on git, no? You should add it as a `buildInput` and then wrap mani to also find it at runtime

teto commented on pkgs/development/tools/mani/default.nix:
```diff
@@ -0,0 +1,44 @@
+    mkdir -p $out/share/{bash-completion/completions,fish/vendor_completions.d,zsh/site-functions}
+    $out/bin/mani completion bash > "$out/share/bash-completion/completions/mani"
+    $out/bin/mani completion fish > $out/share/fish/vendor_completions.d/mani.fish
+    $out/bin/mani completion zsh > $out/share/zsh/site-functions/_mani
```
Though that works, the standard  nixpkgs way uses `installShellCompletion`.

thornycrackers replied:
I figured out why I couldn't get it working before. Updated.

thornycrackers commented:
@teto That's a great suggestion. I added a snippet that I found in another go package called `jump` to generate the autocomplete.

@legendofmiracles  Updates applied from your suggestions.

Thank you both for the review. I am struggling to verify that the changes work correctly though. For the autocomplete I tried `nix-env -f $NIXPKGS -iA mani` but I wasn't able to get the autocomplete to work. To test the git change I tried uninstalling git locally and then adding the `buildInputs` but mani wouldn't actually clone repos unless I specifically had git installed. I'm still fairly new to packaging so your feedback is appreciated.

legendofmiracles commented:
> completion doesn’t work

That’s because your bash is configured to not load completion scripts. It works here where I enabled it in fish.

> git needs to be installed

This patch should fix it: http://0x0.st/-RV-.txt

Another note, your commits and PR titles should follow the contributing guidelines, so: “mani: init at <version>”
and squash the other package related commit

thornycrackers commented:
Updated with the diff and tested mani and it does work now with git not installed, thankyou. I thought that's what `buildInputs` was for? how do I know in the future when to use buildInputs and when to wrap with wrapProgram?

Commits squashed and message updated.

Also updated the completions to use `installShellCompletion` instead of what I copied previously

again, thank you for the review.

legendofmiracles commented on pkgs/development/tools/mani/default.nix:
```diff
@@ -0,0 +1,44 @@
+      --bash <($out/bin/dyff completion bash) \
+      --fish <($out/bin/dyff completion fish) \
+      --zsh <($out/bin/dyff completion zsh)
+    wrapProgram $out/bin/mani \
```
```suggestion
    
    wrapProgram $out/bin/mani \
```
Minor nitpick

thornycrackers replied:
updated.

legendofmiracles reviewed (approved):
> I thought that's what buildInputs was for? how do I know in the future when to use buildInputs and when to wrap with wrapProgram?

You can actually not specify runtime dependencies via nixpkgs (at least not without makeWrapper). What you are specifying are libraries to be linked, and it only differs from nativeBuildInputs (which is often conceived as build dependencies) in a cross compiling sense.

thornycrackers committed:
mani: init at 0.10.0

SuperSandro2000 commented on pkgs/development/tools/mani/default.nix:
```diff
@@ -0,0 +1,45 @@
+
+  # Skip tests
+  # The repo's test folder has a README.md with detailed information. I don't
+  # know how to wrap the dependancies for these integration tests so skip for now.
```
```suggestion
  # know how to wrap the dependencies for these integration tests so skip for now.
```

thornycrackers replied:
updated.
